modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryBufferLimitTest.java

```diff
@@ -0,0 +1,270 @@
+                .setCacheMode(cacheMode)
+                .setBackups(1)
+                .setAffinity(new RendezvousAffinityFunction(false, PARTS)));
+    }
+
+    /** */
+    @Before
+    public void resetMessageCounter() {
+        msgCntr.set(0);
+    }
+
+    /** */
+    @After
+    public void stopAllInstances() {
+        stopAllGrids();
+    }
+
+    /**
+     * @param locBlockPred Block predicate on local node to emulate message delivery issues.
+     * @throws Exception If fails.
+     */
+    private void doTestContinuousQueryPendingBufferLimit(
+        IgniteBiPredicate<ClusterNode, Message> locBlockPred
+    ) throws Exception
+    {
+        ThreadLocalRandom rnd = ThreadLocalRandom.current();
+
+        IgniteEx locIgnite = startGrid(0);
+        IgniteEx rmtIgnite = startGrid(1);
+
+        IgniteCache<Integer, Integer> cache = locIgnite.cache(DEFAULT_CACHE_NAME);
+        CacheConfiguration<Integer, Integer> ccfg = cache.getConfiguration(CacheConfiguration.class);
+
+        for (int i = 0; i < TOTAL_KEYS; i++)
+            cache.put(i, i);
+
+        assertEquals(PARTS, ccfg.getAffinity().partitions());
+
+        AtomicLong lastAcked = new AtomicLong();
+
+        ContinuousQuery<Integer, Integer> cq = new ContinuousQuery<>();
+        cq.setRemoteFilterFactory(FactoryBuilder.factoryOf(RMT_FILTER));
+        cq.setLocalListener((events) ->
+            events.forEach(e ->
+                lastAcked.getAndUpdate(c ->
+                    Math.max(c, ((CacheQueryEntryEvent<?, ?>)e).getPartitionUpdateCounter()))));
+        cq.setLocal(false);
+
+        IgniteInternalFuture<?> updFut = null;
+
+        try (QueryCursor<?> qry = locIgnite.cache(DEFAULT_CACHE_NAME).query(cq)) {
+            awaitPartitionMapExchange();
+
+            for (int j = 0; j < TOTAL_KEYS; j++)
+                putX2Value(cache, rnd.nextInt(TOTAL_KEYS));
+
+            SystemView<ContinuousQueryView> rmtQryView = rmtIgnite.context().systemView().view(CQ_SYS_VIEW);
+            assertEquals(1, rmtQryView.size());
+
+            UUID routineId = rmtQryView.iterator().next().routineId();
+
+            // Partition Id, Update Counter, Continuous Entry.
+            ConcurrentMap<Long, CacheContinuousQueryEntry> pending =
+                getContinuousQueryPendingBuffer(rmtIgnite, routineId, CU.cacheId(DEFAULT_CACHE_NAME), 0);
+
+            spi(locIgnite).blockMessages(locBlockPred);
+
+            updFut = GridTestUtils.runMultiThreadedAsync(() -> {
+                while (!Thread.currentThread().isInterrupted()) {
+                    putX2Value(cache, rnd.nextInt(TOTAL_KEYS));
+                }
+            }, 3, "cq-put-");
+
+            assertNotNull("Partition remote buffers must be inited", pending);
+
+            log.warning("Waiting for pending buffer being overflowed within " + OVERFLOW_TIMEOUT_MS + " ms.");
+
+            boolean await = waitForCondition(() -> pending.size() > PENDING_LIMIT, OVERFLOW_TIMEOUT_MS);
```

Comment:
I think we wait here too long. With 15 secs timeout, it's about 2.5 additional minutes for all tests.

Reply:
Fixed.